lib/improver/nowcasting/pysteps_advection.py

```diff
@@ -0,0 +1,249 @@
+        From the initial cube, generate a precipitation rate array in mm h-1
+        with orographic enhancement subtracted, as required for advection
+
+        Returns:
+            np.ndarray:
+                2D precipitation rate array in mm h-1
+        """
+        self.analysis_cube, = ApplyOrographicEnhancement("subtract").process(
+            self.analysis_cube, self.orogenh)
+        self.analysis_cube.convert_units('mm h-1')
+        return np.ma.filled(self.analysis_cube.data, np.nan)
+
+    def _generate_displacement_array(self, ucube, vcube):
+        """
+        Create displacement array of shape (2 x m x n) required by pysteps
+        algorithm
+
+        Args:
+            ucube (iris.cube.Cube):
+                Cube of x-advection velocities
+            vcube (iris.cube.Cube):
+                Cube of y-advection velocities
+
+        Returns:
+            displacement (np.ndarray):
+                Array of shape (2, m, n) containing the x- and y-components
+                of the m*n displacement field (format required for pysteps
+                extrapolation algorithm)
+        """
+        def _calculate_displacement(cube, interval, gridlength):
+            """
+            Calculate displacement for each time step using velocity cube and
+            time interval
+
+            Args:
+                cube (iris.cube.Cube):
+                    Cube of velocities in the x or y direction
+                interval (int):
+                    Lead time interval, in minutes
+                gridlength (float):
+                    Size of grid square, in metres
+
+            Returns:
+                np.ndarray:
+                    Array of displacements in grid squares per time step
+            """
+            cube_ms = cube.copy()
+            cube_ms.convert_units('m s-1')
+            displacement = cube_ms.data*interval*60. / gridlength
+            return np.ma.filled(displacement, np.nan)
+
+        gridlength = calculate_grid_spacing(self.analysis_cube, 'metres')
+        udisp = _calculate_displacement(ucube, self.interval, gridlength)
+        vdisp = _calculate_displacement(vcube, self.interval, gridlength)
+        displacement = np.array([udisp, vdisp])
+        return displacement
+
+    def _reformat_analysis_cube(self):
+        """
+        Add forecast reference time and forecast period coordinates (if they do
+        not already exist) and nowcast attributes to analysis cube
+        """
+        coords = [coord.name() for coord in self.analysis_cube.coords()]
+        if "forecast_reference_time" not in coords:
+            frt_coord = self.analysis_cube.coord('time').copy()
+            frt_coord.rename('forecast_reference_time')
+            self.analysis_cube.add_aux_coord(frt_coord)
+        if "forecast_period" not in coords:
+            self.analysis_cube.add_aux_coord(
+                AuxCoord(np.array([0], dtype=np.int32),
+                         'forecast_period', 'seconds'))
```

Comment:
Why not use existing function?
improver.utilities.temporal.forecast_period_coord(self.analysis_cube)

Reply:
Usually I'd call an external utility in order to reduce the complexity of my code (eg to get rid of 5-6 lines of code and replace it with one line), at the potential cost of making the user have to dig a bit more to find out what's going on.  In this case, calling `forecast_period_coord` doesn't reduce the complexity of my code, which is a one-line call to an iris function; but does incur the cost of dependency and digging.  So I used the iris function.

Reply:
From a maintenance perspective, when we next decide to change the forecast_period meta-data standard, using the central function will reduce the overhead of that maintenance. If you don't think this argument is strong enough though, then I don't mind sticking with what you've put in.

Reply:
I'm not happy with what `forecast_period_coord` does and I don't think it's suitable for this case.  If I were to use it, I'd want to refactor the utility to deal with cube lists.  Otherwise, I think the added complexity is more costly than the maintenance cost for changing metadata standards.

Reply:
Ok. We'll stick with the PR as it stands.